src/gnetcli_adapter/gnetcli_adapter.py

```diff
@@ -396,84 +432,133 @@ async def deploy(self, device: Device, cmds: CommandList, args: DeployOptions, p
             ip=ip,
         )
         do_reload: bool = args.entire_reload.value == "yes"
-        seen_exc: list[Exception] = []
-        reload_cmds: dict[str, CommandList] = {}
-        total_cmds = 0
         if isinstance(cmds, dict): # PC
-            for file, cmd in cmds["cmds"].items():
-                if isinstance(cmd, bytes):
-                    cmd = cmd.decode()
-                reload_cmds[file] = CommandList([Command(cmd, suppress_nonzero=True) for cmd in cmd.splitlines()])
-                total_cmds += len(reload_cmds[file])
+            reload_cmds = self._get_reload_cmds(cmds)
+            files = self._get_files(cmds)
             run_cmds = CommandList()
-            files: Dict[str, File] = {file: File(content=content, status=None) for file, content in cmds["files"].items()}
-            await self.api.upload(hostname=device.fqdn, files=files, host_params=host_params)
         else:
-            total_cmds = len(cmds)
             run_cmds = cmds
+            files = {}
+            reload_cmds = {}
 
-        done_cmds = 0
-        async with self.api.cmd_session(hostname=device.fqdn) as sess:
-            result: List[pb.CMDResult] = []
-            for cmd in run_cmds:
-                if progress_bar:
-                    progress_bar.set_progress(device.fqdn, done_cmds, total_cmds, suffix=cmd.cmd)
+        with self._init_progress_tracker(device, progress_bar) as tracker:
+            tracker.set_total(self._get_total(run_cmds, reload_cmds, files))
+            if files:
+                tracker.upload_files(list(files))
+                await self.api.upload(hostname=device.fqdn, files=files, host_params=host_params)
+                tracker.files_uploaded()
+
+            async with self.api.cmd_session(hostname=device.fqdn) as sess:
+                try:
+                    seen_exc, res = await self._deploy_cmds(
```

Comment:
We can still throw exception and left tracker without proper finish call, for example in api.upload() or in api.cmd_session(). What about something like this:
```python
with self._init_progress_tracker(device, progress_bar) as tracker:
    try:
        await _deploy(...)
    except Exception as e:
        tracker.finish_err(f"Unexpcted exception {e}")
```